File: 270.closest-binary-search-tree-value.java
/*
 * @lc app=leetcode id=270 lang=java
 *
 * [270] Closest Binary Search Tree Value
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */

class Solution {
    int res;
    double target;

    public int closestValue(TreeNode root, double target) {
        this.res = root.val;
        this.target = target;
        dfs(root);

        return res;

    }

    public void dfs(TreeNode root) {
        if (root == null)
            return;

        if (Math.abs((double) root.val - target) < Math.abs((double) res - target)) {
            res = root.val;
        }

        dfs(root.left);
        dfs(root.right);
    }
}

// @lc code=end

File: 270.closest-binary-search-tree-value.py
#
# @lc app=leetcode id=270 lang=python3
#
# [270] Closest Binary Search Tree Value
#


# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def closestValue(self, root: Optional[TreeNode], target: float) -> int:
        #  number of nodes in the tree is in the range [1, 10^4]
        res = root.val

        while root:
            if abs(target - root.val) < abs(target - res):
                res = root.val
            # the equal situation
            if abs(target - root.val) == abs(target - res):
                res = min(res, root.val)

            if root.val > target:
                root = root.left
            else:
                root = root.right

        return res


# @lc code=end
